pandas/core/groupby/generic.py

```diff
@@ -322,29 +332,109 @@ def _aggregate_multiple_funcs(self, arg, _level):
+        assert len(output) == len(columns)
+        assert list(output.keys()) == sorted(output.keys())
+
+        result: Union[Series, DataFrame]
+        if len(output) > 1:
+            result = DataFrame(output, index=index)
+            result.columns = columns
         else:
-            name = self._selection_name
-            if name is None:
-                name = self._selected_obj.name
-            return Series(output, index=index, name=name)
+            result = Series(output[0], index=index, name=columns[0])
+
+        return result
+
+    def _wrap_aggregated_output(
+        self, output: Mapping[int, Union[Series, np.ndarray]], columns: Index
+    ) -> Union[Series, DataFrame]:
+        """
+        Wraps the output of a SeriesGroupBy aggregation into the expected result.
+
+        Parameters
+        ----------
+        output : Mapping[int, Union[Series, np.ndarray]]
+            Dict where the key represents the columnar-index and the values are
+            the actual results.
+        columns : pd.Index
+            Columns to apply to the output.
+
+        Returns
+        -------
+        Series or DataFrame
+
+        Notes
+        -----
+        In the vast majority of cases output and columns will only contain one
+        element. The exception is operations that expand dimensions, like ohlc.
+        """
+        assert list(output.keys()) == sorted(output.keys())
 
-    def _wrap_aggregated_output(self, output, names=None):
         result = self._wrap_series_output(
-            output=output, index=self.grouper.result_index, names=names
+            output=output, index=self.grouper.result_index, columns=columns
         )
         return self._reindex_output(result)._convert(datetime=True)
 
-    def _wrap_transformed_output(self, output, names=None):
-        return self._wrap_series_output(
-            output=output, index=self.obj.index, names=names
+    def _wrap_transformed_output(
+        self, output: Mapping[int, Union[Series, np.ndarray]], columns: Index
+    ) -> Series:
+        """
+        Wraps the output of a SeriesGroupBy aggregation into the expected result.
+
+        Parameters
+        ----------
+        output : dict[int, Union[Series, np.ndarray]]
+            Dict with a sole key of 0 and a value of the result values.
+        columns : pd.Index
+            Columns to apply to the output.
+
+        Returns
+        -------
+        Series
+
+        Notes
+        -----
+        output and columns should only contain one element. These are containers
+        for generic compatability with the DataFrameGroupBy class.
+        """
+        assert list(output.keys()) == sorted(output.keys())
+
+        result = self._wrap_series_output(
+            output=output, index=self.obj.index, columns=columns
         )
 
+        # No transformations increase the ndim of the result
+        # Unfortunately need to cast for mypy to know this
+        result = cast(Series, result)
```

Comment:
do we have a policy on cast vs assert?  i think i prefer assert

Reply:
> do we have a policy on cast vs assert?

should we have a policy?

when adding either, both are doing effectively the same thing in silencing mypy.

Mypy is alerting us to a potential issue. and our tests won't be covering whatever issue mypy is alerting us to. so an assert won't add anything in this case.

one difference is the failure mode if the alerted issue is a real issue. we'll either get an AssertionError or an error on the line with the issue when using a cast.

The cast is a noop at runtime unlike assert. assert something is not None is probably OK not so sure about isinstance checks. 

Redundant casts can be identified with `--warn-redundant-casts`. not sure how to identify unused asserts and track the ones added for the benefit of static type checking.


> i think i prefer assert

so casts are probably preferable, but the asserts are maybe clearer to the reader.

we can use `not` with asserts, so it is cleaner when trying to exclude a type and is an advantage over cast.

Reply:
for this particular scenario, since we don't have a check in this method that the `output` argument passed has len == 1, then an assert could be more appropriate here?

Reply:
Changed to assert